src/solvers/refinement/string_constraint_generator_valueof.cpp

```diff
@@ -514,56 +513,60 @@ exprt string_constraint_generatort::add_axioms_for_correct_number_format(
 }
 
 /// add axioms corresponding to the Integer.parseInt java function
-/// \par parameters: function application with one string expression
+/// \param f: a function application with either one string expression or one
+///   string expression and an expression for the radix
 /// \return an integer expression
 exprt string_constraint_generatort::add_axioms_for_parse_int(
   const function_application_exprt &f)
 {
-  string_exprt str=get_string_expr(args(f, 1)[0]);
+  PRECONDITION(f.arguments().size()==1 || f.arguments().size()==2);
+  string_exprt str=get_string_expr(f.arguments()[0]);
+  const exprt radix=
+    f.arguments().size()==1?from_integer(10, f.type()):f.arguments()[1];
+
   const typet &type=f.type();
   symbol_exprt i=fresh_symbol("parsed_int", type);
   const refined_string_typet &ref_type=to_refined_string_type(str.type());
   const typet &char_type=ref_type.get_char_type();
-  exprt zero_char=constant_char('0', char_type);
   exprt minus_char=constant_char('-', char_type);
   exprt plus_char=constant_char('+', char_type);
   assert(type.id()==ID_signedbv);
-  exprt ten=from_integer(10, type);
 
   exprt chr=str[0];
   exprt starts_with_minus=equal_exprt(chr, minus_char);
   exprt starts_with_plus=equal_exprt(chr, plus_char);
-  exprt starts_with_digit=binary_relation_exprt(chr, ID_ge, zero_char);
+  exprt starts_with_digit=
+    not_exprt(or_exprt(starts_with_minus, starts_with_plus));
 
   // TODO: we should throw an exception when this does not hold:
-  exprt correct=add_axioms_for_correct_number_format(str);
+  exprt correct=add_axioms_for_correct_number_format(str, radix);
   axioms.push_back(correct);
 
   for(unsigned size=1; size<=10; size++)
```

Comment:
I think there can be problem with the size here: 10 may be actual limit for for integer written in decimal, but this may vary according to the radix (for instance it should be 9 for hexadecimal `-7FFFFFFF` and 32 for binary), so this code may be to constrained for radix smaller than 10.
I don't think you need to fix that in this pull request as this already improve the current implementation, but we should have a `TODO` note about that,  and maybe raise an issue.